runtime/fs_fake.js

```diff
@@ -62,15 +63,64 @@ MlFakeDevice.prototype.exists = function(name) {
   this.lookup(name);
   return this.content[name]?1:0;
 }
-MlFakeDevice.prototype.mkdir = function(name,mode) {
+MlFakeDevice.prototype.mkdir = function(name,mode, raise_unix) {
+  var unix_error = raise_unix && caml_named_value('Unix.Unix_error');
+  if(this.exists(name)) {
+    if (unix_error) {
+      caml_raise_with_args(unix_error, [8, caml_string_of_jsstring("mkdir"), caml_string_of_jsstring(this.nm(name))]);
+    }
+    else {
+      caml_raise_sys_error(name + ": File exists");
+    }
+  }
+  var parent = /^(.*)\/[^/]+/.exec(name);
+  parent = (parent && parent[1]) || '';
+  if(!this.exists(parent)){
+    if (unix_error) {
+      caml_raise_with_args(unix_error, [20, caml_string_of_jsstring("mkdir"), caml_string_of_jsstring(this.nm(parent))]);
+    }
+    else {
+      caml_raise_sys_error(parent + ": No such file or directory");
+    }
+  }
+  if(!this.is_dir(parent)){
+    if (unix_error) {
+      caml_raise_with_args(unix_error, [26, caml_string_of_jsstring("mkdir"), caml_string_of_jsstring(this.nm(parent))]);
+    }
+    else {
+      caml_raise_sys_error(parent + ": Not a directory");
+    }
+  }
   this.create_dir_if_needed(this.slash(name));
 }
-MlFakeDevice.prototype.rmdir = function(name) {
+MlFakeDevice.prototype.rmdir = function(name, raise_unix) {
+  var unix_error = raise_unix && caml_named_value('Unix.Unix_error');
   var name_slash = (name == "")?"":(this.slash(name));
   var r = new RegExp("^" + name_slash + "([^/]+)");
+  if(!this.exists(name)) {
+    if (unix_error) {
```

Comment:
This error handling here is very repetitive, can we do something about it ?

Reply:
At first glance, I don't see a good way to reduce the repetition. Any helper methods we add for this will require many arguments to handle the variation between different syscalls ("mkdir" and "rmdir" currently). We could implement them to take JS objects like `{ message, code, syscall, path }` and make a helper similar to the `raise_nodejs_error` helper in fs_node.

I'll keep thinking about this.